src/app/CommandSender.h

```diff
@@ -36,15 +36,69 @@
 #include <system/SystemPacketBuffer.h>
 
 #include <app/Command.h>
+#include <app/MessageDef/CommandPath.h>
+#include <app/MessageDef/StatusElement.h>
 
 #define COMMON_STATUS_SUCCESS 0
 
 namespace chip {
 namespace app {
 
-class CommandSender : public Command, public Messaging::ExchangeDelegate
+class CommandSender final : public Command, public Messaging::ExchangeDelegate
 {
 public:
+    // TODO: Move various CommandResponse related callbacks to this delegate.
+    class Delegate
+    {
+    public:
+        virtual ~Delegate() = default;
+
+        /**
+         * OnResponse will be called when the response from server arrived.
+         *  - When StatusElement is received, aStatus will be filled with the content of the StatusElement, aData will be a
```

Comment:
I'm not certain this is the API pattern we want.

In most parts of the code-base, there are usually pairs of callbacks defined for any given fail-able operation: a success callback and a failure callback.

Receipt of non-empty data, empty-data, or a status element with success in it should all be dispatched through this `OnResponse` callback. status elements with failure codes I don't think should be routed through this callback, since it really makes it inconsistent with the patterns in the rest of the code-base.

Reply:
I'm also thinking about this, in my mind, the result can be 
Success, Expected Error and Unexpected Error
The Expected Error can be interaction model protocol code (i.e. Command not found, then server replied with the state code)
And Command specific error codes, we cannot deliver this to OnError since the actual type varies from different commands / clusters.

The unexpected error is the server size misbehaved, e.g. network error, / malformed response (bad TLV) etc.

I'm not certain whether a client will expect a failure status code in im when sending commands, if this is not true, then we should put im error with protocol error handling into seperate callback.

I'm also thinking about some other behaviors, i.e. HTTP requests, seems most library will regard HTTP error as exceptions, so maybe we should call OnError on interaction model errors.